Allow us time_t timespec to be 64bit for POSIX timedWait

ocrete opened the pull request:
#### f63657fbf8c1518df18432a50486cbc2f2e8ce4c
<pre>
Allow us time_t timespec to be 64bit for POSIX timedWait
<a href="https://bugs.webkit.org/show_bug.cgi?id=244302">https://bugs.webkit.org/show_bug.cgi?id=244302</a>

The current code assumes that the fields in timespec are 32 bit. This
causes timedWait() to degenerate into wait() if the system clock is
set beyond year 2038. Instead allow the full size of time_t,  which is
64-bit on anything respecting POSIX 2001 or C11, and then replace the
validation to use std::numeric_limit&lt;time_t&gt;::max()

Reviewed by Adrian Perez de Castro

* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::ThreadCondition::timedWait):

Canonical link: <a href="https://commits.webkit.org/253858@main">https://commits.webkit.org/253858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0dcad5402c136b48e377a66c2f73b5867d9d5252

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96351 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149826 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29702 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79362 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91257 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92866 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23999 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74052 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23863 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78973 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79211 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66878 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79111 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27424 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13047 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72787 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27372 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14062 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25929 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2718 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29056 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36920 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75603 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33339 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16743 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
```diff
@@ -609,15 +609,17 @@ bool ThreadCondition::timedWait(Mutex& mutex, WallTime absoluteTime)
-    if (absoluteTime > WallTime::fromRawSeconds(INT_MAX)) {
+    const time_t timetMax = sizeof(time_t) == 4 ? INT_MAX : INT64_MAX;
+
+    if (absoluteTime > WallTime::fromRawSeconds(timetMax)) {
```
Is there any reason to not use `std::numeric_limits<time_t>` here?

The following should do:

```suggestion
    if (absoluteTime > WallTime::fromRawSeconds(std::numeric_limits<time_t>::max())) {
```

ocrete replied:
The only reason is that my C++ knowledge is quite limited, will update the branch !

philn replied:
Source of a new clang warning:
```
/app/webkit/Source/WTF/wtf/posix/ThreadingPOSIX.cpp:612:49: warning: implicit conversion from 'long' to 'double' changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
    if (absoluteTime > WallTime::fromRawSeconds(std::numeric_limits<time_t>::max())) {
                       ~~~~~~~~                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

aperezdc reviewed (changes requested):
Thanks for the patch! I do have a comment, please take a look.

ocrete committed:
Allow us time_t timespec to be 64bit for POSIX timedWait
https://bugs.webkit.org/show_bug.cgi?id=244302

The current code assumes that the fields in timespec are 32 bit. This
causes timedWait() to degenerate into wait() if the system clock is
set beyond year 2038. Instead allow the full size of time_t,  which is
64-bit on anything respecting POSIX 2001 or C11, and then replace the
validation to use std::numeric_limit<time_t>::max()

Reviewed by Adrian Perez de Castro

* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::ThreadCondition::timedWait):

Canonical link: https://commits.webkit.org/253858@main
